ICM lifecycle and API compatibility governance - for review (#545)
camaraproject/ReleaseManagement: documentation/SupportingDocuments/icm-lifecycle-and-compatibility.md

- **ICM deployment info**: information defined by an ICM version that applies to API deployments (including at API Provider/Consumer design time and at runtime) — auth flows, grant types, assertion format and lifetime, token processing, claim handling, and similar behaviors agreed between an API Provider and API Consumer. API-Provider-governed in their contractual relationship with API Consumers.
- **API version ICM-compatibility**: the guarantee that an API version's definition respects the ICM design info of a given ICM version. A property of the API definition relative to an ICM version declared in its `x-camara-min-icm` field.
- **API deployment ICM-compatibility**: the guarantee that an API Provider's or API Consumer's deployment (a) deploys ICM-compatible API versions and (b) implements the ICM deployment info of the ICM version it claims to operate under.
- **ICM-compatibility** (umbrella): both API version and API deployment ICM-compatibility together. Successful integration of an API Consumer with an API Provider requires both to hold.

> - **ICM version**: a Semantic Versioning (SemVer 2.0) compliant version number with major, minor and patch components, identifying a specific set of ICM artefacts. Starting with ICM 1.0.0, major-version increments indicate breaking changes for API definitions, API deployments, or both — changes that cannot be expressed additively.

Reviewer: `a Semantic Versioning (SemVer 2.0) compliant version number with major, minor and patch components,`

a Semantic Versioning (SemVer 2.0) compliant version number with major, minor, and patch components,

- **ICM lifecycle states**: Supported / Deprecated / Retired / Revoked are the four possible lifecycle states of an ICM version (see §5). Lifecycle states apply per ICM version.
- **ICM governance**: the decision process to transition an ICM version to a different lifecycle state (see §5).
- **Exception (waiver)**: a time-bound, governance-approved authorization that permits a specific (API version, ICM version) pair outside the normal ICM-compatibility guarantee.
- **Compatibility matrix**: the derived artifact listing which (API version, ICM version) pairs are ICM-compatible at a given point in time (see §9).

## 3. ICM-compatibility

ICM-compatibility concerns two distinct classes of information defined by an ICM version, referred to as ICM design info and ICM deployment info. Each class has its own scope, governance, and announcement mechanism, and either or both can be changed in a given ICM version independently of the other.

### 3.1 API version ICM-compatibility (design-time)

> **An API version is ICM-compatible with an ICM version when its definition respects the ICM design info of that ICM version** — its scope format, `securitySchemes` shape, schemas, operations, and `info.description` text use constructs and conventions defined by the ICM version.

Reviewer: `its scope format, `securitySchemes` shape, schemas, operations, and `info.description` text use constructs and conventions defined by the ICM version`

Its scope format, securitySchemes shape, schemas, operations, and info.description text constructs and conventions, as defined by the ICM version

### 3.2 API deployment ICM-compatibility (runtime)

**An API Provider's or API Consumer's deployment is ICM-compatible with an ICM version when (a) it deploys ICM-compatible API versions, and (b) the runtime interactions between API Consumer and API Provider implement the ICM deployment info of that ICM version** — auth flows, grant types, assertion format and lifetime, token processing, and claim handling.

- **Owned and governed by the API Provider.** Deployment ICM-compatibility is a contractual matter between API Provider and API Consumer, communicated by the Provider through onboarding documentation, Provider metadata (for example via OIDC discovery), and similar mechanisms. The mandatory `info.description` text in every CAMARA API spec already states that "the specific authorization flows to be used will be agreed upon during the onboarding process."
- **Identified at deployment/runtime.** An API Consumer determines the applicable ICM version's deployment info by inspecting API Provider metadata and onboarding artifacts; the API version alone does not pin a specific ICM version on the deployment side.

> - **Not recorded in the CAMARA compatibility matrix.** The matrix governs API version ICM-compatibility (design-time). API deployment ICM-compatibility is the API Provider's responsibility.

Reviewer 2: As discussed at the last ICM meeting (https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/843055110/2026-06-03+ICM+Minutes), this section (3.2) could be a little misleading, as the reader may understand that the ICM version to be used by the API Consumer and the API Provider can be discovered at runtime, which is not currently the case. This specification relies on the same principle written in the API Specs info.description template, and the ICM version to be used will be determined during the onboarding process. Perhaps it would suffice to remove the 'runtime' references here.

CC @tanjadegroot @hdamker

Author: I have reworded the section 3.2, please check if OK.

Reviewer 2: The changes look good to me.

> ### 3.3 Maintaining ICM-compatibility

Reviewer: Could you add a diagram or a process flow to illustrate how the different stakeholders (API designers, API Providers, API Consumers) interact to maintain ICM-compatibility ?

Author: yes, I will give this a thought and will come back on it

Author: Added the table with requested columns in section 3.3 columns - see commit: https://github.com/camaraproject/ReleaseManagement/pull/545/changes/2048e5a51bc1042048cd7c1c1e53abec056ce45b

- **A change to ICM deployment info** (for example, tightening assertion lifetime) → API Provider deployments and API Consumer implementations must update → does NOT by itself imply a new API version. The API definition is unchanged; only the runtime behavior changes.
- **An ICM change to both** → both consequences apply.

### 3.4 Path from ICM through Commonalities into API definitions

ICM design info changes do not reach API definitions directly. The path is:

1. ICM defines or updates design info (security schemas, scope format, mandatory text, etc.).
2. Commonalities updates the API Design Guide (and other relevant artifacts) to mandate the new design info for API definitions.
3. API Sub Projects update their API versions to align with the new Commonalities API Design Guide.

This means an ICM design info change typically also triggers a Commonalities update. How tightly ICM and Commonalities lifecycles must be coupled is an open item (see §12).

> ## 4. ICM versioning

Reviewer: Could you clarify how changes in ICM deployment info are versioned and how they interact with changes in ICM design info. It would be beneficial to add a subsection or a note to explain the versioning process for ICM deployment info changes and their impact on ICM design info.

Author: The ICM version reflects changes in both types of ICM info at the same time.
ICM design info and API deployment info are not versioned separately.

A change in ICM deployment info MAY (1) or MAY NOT (2) lead to a change of API design info, e.g.

1. if a new ICM deployment info needs to be captured in API specifications, such as a new auth flow, introducing a new type of credential/token.

2. a breaking change in the ICM deployment info such as the introduction of a 300s client-assertion lifetime cap has no impact on the API design info, but will lead to a major ICM version change. 

Note: I am not an ICM specialist, so the above examples could be wrong. Hence, in open issue nr 10 ([Example content for the §5.4.2 release-note tables](https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/822444052/ICM+lifecycle+-+open+issues+WIP#Example-content-for-the-%C2%A75.4.2-release-note-tables)), we are asking the ICM team to provide more examplse of such changes. e.g. taken from previous ICM versions.

Reviewer 2: I added some examples to https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/822444052/ICM+lifecycle+-+open+issues+WIP#Example-content-for-the-%C2%A75.4.2-release-note-tables and suggested changes to the table columns to make them as actionable as possible.

Author: @sebdewet Seb's comment is adressed by this [commit](https://github.com/camaraproject/ReleaseManagement/pull/545/changes/bf43e23125297945f079822466ec8f3d05bb3816). Please check.

Author: > I added some examples to https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/822444052/ICM+lifecycle+-+open+issues+WIP#Example-content-for-the-%C2%A75.4.2-release-note-tables and suggested changes to the table columns to make them as actionable as possible.

@jpengar The table updates have been integrated through this [commit](https://github.com/camaraproject/ReleaseManagement/pull/545/changes/e529a33a1b91aaf88563b230c605bb2f79a39aea)

| Duration of Deprecated state | 12 months | Active migration period for API deployments before a Deprecated ICM version is Retired; impacts API version ICM-compatibility |
| Concurrent support requirement by API deployments | API Providers shall continue to deploy the most recent previous Supported major ICM version next to the latest published Supported major ICM version (see period defined above) | Applies to ICM-compatible API deployments during this period. |
| Exceptions | Conditions permitting governed ICM lifecycle state transition  | Explicit and recorded governance action per exception; see §10. |

These values are starting points for WG discussion. <!-- to be removed when WG agrees -->

### 5.4 ICM version - Release notes

#### 5.4.1 Publication of lifecycle state

The lifecycle state is published in each ICM version's release notes, as a table in the release notes template. No separate governance artifact is required. Each ICM version release carries the lifecycle state for all ICM versions. State transitions are committed at ICM public release unless an out-of-cycle governance action specifies otherwise.

> The published lifecycle state must be available in machine-readable form for the CAMARA validation support to consume (§6.5). Until a schema is defined, the ICM version release notes are the single authoritative source — readable by humans but not by tooling.

Reviewer 2: As discussed at the last ICM meeting  (https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/843055110/2026-06-03+ICM+Minutes), it was agreed that we would need a template for ICM release notes so that the CAMARA tooling could automatically process the ICM lifecycle state. @hdamker said that we could provide a format proposal, but to do so, we need the data and information to be included.

As far as I understand, this information should include the ICM lifecycle state, for which there is already table proposal from @tanjadegroot at https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/822444052/ICM+lifecycle+-+open+issues+WIP#Example-ICM-lifecycle-state-table-in-ICM-release-notes, which looks good to me. What do you think?

My only question is whether the proposed table should somehow consider the 'per-version overrides' mentioned in the document.

Author: yes you are right - will tweek the wiki page example a bit more to include that. information.

Author: The lifecycle state table example has been added below, including the override example. (Same on wiki page). see commit: https://github.com/camaraproject/ReleaseManagement/pull/545/changes/72739f8fe38c0fc2259837ee48d09c3724f03df6

| ICM version | Breaking design info change | Affected construct | Impact on API definition |
|---|---|---|---|
| _2.0.0_ | _introduces new scope format X_ | _where it appears in an API definition_ | _what API teams need to do_ |

**Table B — Breaking ICM deployment info changes** (impacting API deployments; see §3.2). Used by API Providers and API Consumers to plan deployment updates.

| ICM version | Breaking deployment info change | Affected runtime behavior | Impact on API deployment |
|---|---|---|---|
| _0.3.0_ | _introduces 300s client-assertion lifetime cap_ | _which deployment behavior_ | _what Provider and Consumer need to do_ |

The exact column layout is subject to refinement once concrete examples are filled in (see §12).

## 6. API version ICM-compatibility - details

### 6.1 Minimum ICM version

An API version declares its minimum ICM version via the `x-camara-min-icm` field in its API definition:

```yaml
info:
  # The version of the CAMARA Identity and Consent Management (ICM)
  # specification that this API definition is built against.
  # This API version is compatible with higher ICM versions with the same major version number.
  x-camara-min-icm: 1.2.0
```

> This declared ICM version is the lowest ICM version that the API definition relies on and which is Supported at the time of public release of this API version.

Reviewer: `This declared ICM version is the lowest ICM version that the API definition relies on and which is Supported at the time of public release of this API version`

This declared ICM version is the lowest ICM version on which the API definition relies and which is Supported at the time of the public release of this API version.

### 6.3 API designer responsibilities

Per §3.3, an API designer must release an update of the API version when it is no longer ICM-compatible. This incompatibility may be due to:

- Lifecycle state change of the referenced ICM version to Retired or Revoked. API designers must release a new API version referencing a Supported ICM version.
- Availability of a new major ICM version with changes that break either or both:
  - API version ICM-compatibility — affecting ICM design info (scope format, securitySchemes syntax, schemas, operations): API designers must release a new API version compatible with the new major ICM version.
  - API deployment ICM-compatibility — affecting ICM deployment info (auth flows, grant types, assertion format and lifetime, token processing): No need to release a new API version. Maintaining API deployment ICM-compatibility is the responsibility of API Providers and API Consumers.

### 6.4 Determining the `x-camara-min-icm` value

> At each API version public release (new major, minor, or patch), the API Sub Project determines the value using the following formula:

Reviewer: Could you please add a concrete example to illustrate how to apply the formula.

Author: Examples added, please check if OK.

Reviewer: ok for me

1. Implement a Supported ICM version.
2. When a new major ICM version is published and is in Supported state, plan implementation of that ICM version. During the period in which the previous major ICM version remains in the Supported state, keep that major ICM version running in parallel, so that API versions targeting either Supported major ICM version remain deployable while API Providers and Consumers plan migration of impacted API versions.
3. Announce the ICM versions they implement, in a form that API Consumers can discover (for example via Provider metadata or onboarding documentation).
4. Publish an **ICM-compatibility statement** covering:
   - **API version ICM-compatibility**: for each API version offered.
   - **API deployment ICM-compatibility**: the ICM versions implemented.

This statement declares the API Provider's ICM-compatibility baseline.

The mechanism by which an API Provider provides multiple major ICM versions concurrently is an implementation choice and is not prescribed by this guideline; what matters is that each implemented ICM version is clearly announced by the API Provider for use by API Consumers during the relevant migration period.

> An API Provider cannot declare a newly deployed API version as ICM-compatible when offering only a Deprecated or Retired ICM version, even if the API version would technically work with those older ICM versions.

Reviewer: `An API Provider cannot declare a newly deployed API version as ICM-compatible when offering only a Deprecated or Retired ICM version, even if the API version would technically work with those older ICM versions.`

An API Provider cannot declare a newly deployed API version as ICM-compatible if it offers only a Deprecated or Retired ICM version, even if the API version would technically work with those older ICM versions.

Author: Changed to "if their implementation provides only  ..."

    AND ICM vY is in state {Supported, Deprecated} )
  OR a governance decision extends ICM-compatibility to (API vX, ICM vY.major)
  OR a governance approved, time-bound ICM-compatibility exception exists for (API vX, ICM vY)
```

The lifecycle state condition applies to the specific ICM version `vY`, inherited from its major ICM version, unless a per-version override is in place. 

ICM versions in the Retired or Revoked state are excluded from the ICM-compatibility matrix, unless an approved exception has been decided.

The matrix of ICM-compatible pairs (API version, ICM version) are the CAMARA-supported combinations that API Providers may offer and API Consumers may consume.

### 9.2 ICM-compatibility matrix - updates

ICM-compatibility matrix updates are triggered by:
- New ICM version public release (lifecycle state of other ICM versions may change).
- New API version public release (new row added to the matrix; existing rows unchanged).
- Exception granted, modified, or expired.

### 9.3 ICM-compatibility matrix - ownership

The ICM-compatibility matrix is published by Release Management. It is computed, not hand-edited. Only exceptions require human governance action.

> ## 10. Exception mechanism

Reviewer: Could you add an example of an exception document to clarify the process of requesting and approving exceptions ?

Author: I have added examples of exception decision record, and realize we were missing the lifecyce state transition exception (and record in this section as well) 

Please check wiki proposal as a start [here](https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/822444052/ICM+lifecycle+-+open+issues+WIP#Exception-grant-process)

it indicates owners, but not yet the full process

Author: Updated the Exceptions section: see https://github.com/camaraproject/ReleaseManagement/pull/545/changes/a27f56b784b2dc39626da11a296604d302042c6b

> # ICM Lifecycle and API Compatibility Governance

Reviewer 2: I think it would be a good idea to add an index to the document.

Author: Done

# ICM Lifecycle and API Compatibility Governance

**Version:** Draft 3 (2026-07-29)

> **Status:** Draft for ICM WG discussion / approval; incorporates V2 working merge and ICM team review feedback.

Reviewer 2: Should WG be replaced by 'Working Group' throughout the document? Should we create an acronyms section for terms such as WG, OAS, ICM and TSC?

Author: - the lines where "WG" appears should eventually be removed from the document. 
- ICM is introduced in the first occurence and should be obvious from the CAMARA context, but we can add a link to the repo.
- OAS - expanded first occurrence in flossary section + link (btw Commonalities is referencing swagger rather than the official spec website) 
- TSC - removed from section 8 (just "governance" is enough); added abbreviation in firts occurrence in section on exceptions.

not documented in the glossary, but in other Commonalities and Release Management documentation (API, meta-release, semver, scope, etc) so changed the reference from "Glossary", to links to the wider project documentation.

So in the end adding an acronym section in this document is not needed IMHO, but happy to do so if you relly want it :-)

Reviewer 2: It's OK, you can ignore this comment then.

Author: the above is addressed in https://github.com/camaraproject/ReleaseManagement/commit/2d2bcbc11f7b3c756c9fbe1fe7ea42de61fe489c

**Scope:** Response to the ICM WG request to Release Management (per [ICM#324](https://github.com/camaraproject/IdentityAndConsentManagement/issues/324), [ICM#340](https://github.com/camaraproject/IdentityAndConsentManagement/issues/340), [ReleaseManagement#351](https://github.com/camaraproject/ReleaseManagement/issues/351)) to define governance for ICM version evolution and its dependencies with CAMARA API versions.

---

# Table of Contents

1. [Scope and Purpose](#1-scope-and-purpose)
2. [Glossary](#2-glossary)

> 

Reviewer 2: Should we add a References section, consolidating links to the Commonalties API Design Guide and the ICM Specification Repository? Currently, links only appear inline where they are first mentioned.

Author: I have added the list of references in section 2 and renamed the section 2 to "Glossary and References. 
I kept the inline refs as they are slightly different in some cases.

   5.3 [Duration of ICM lifecycle states](#53-duration-of-icm-lifecycle-states)  
   5.4 [ICM version – Release notes](#54-icm-version---release-notes)
    
      5.4.1 [Publication of lifecycle state](#541-publication-of-lifecycle-state)  
      5.4.2 [ICM version change tables](#542-icm-version-change-tables)

6. [API version ICM-compatibility – details](#6-api-version-icm-compatibility---details)
   
   6.1 [Minimum ICM version](#61-minimum-icm-version)  
   6.2 [ICM governance responsibilities](#62-icm-governance-responsibilities)  
   6.3 [API designer responsibilities](#63-api-designer-responsibilities)  
   6.4 [Determining the x-camara-min-icm value](#64-determining-the-x-camara-min-icm-value)  
   6.5 [CAMARA validation support](#65-camara-validation-support)

7. [API deployment ICM-compatibility – details](#7-api-deployment-icm-compatibility---details)
    
   7.1 [API Provider responsibilities](#71-api-provider-responsibilities)  
   7.2 [API Consumer responsibilities](#72-api-consumer-responsibilities)

8. [ICM version release cadence](#8-icm-version-release-cadence)
    
   8.1 [Signal/Sync meta-release alignment](#81-signalsync-meta-release-alignment)  

>    8.2 [Out-of-cycle ICM releases](#82-out-of-cycle-icm-releases)

Reviewer 2: Missing Section 8.3

Author: fixed !

Reviewer 2: It seems that applying the fix broke the numbering of the subsequent sections.
<img width="1295" height="526" alt="image" src="https://github.com/user-attachments/assets/da8ecd12-4571-4fa1-879b-09dc5bf23a46" />

Author: fixed

9. [ICM-compatibility matrix](#9-icm-compatibility-matrix)
   
   9.1 [ICM-compatibility matrix derivation](#91-icm-compatibility-matrix-derivation)  
   9.2 [ICM-compatibility matrix – updates](#92-icm-compatibility-matrix---updates)  
   9.3 [ICM-compatibility matrix – ownership](#93-icm-compatibility-matrix---ownership)

10. [Exception mechanism](#10-exception-mechanism)

11. [ICM 0.x handling](#11-icm-0x-handling)

12. [Open governance points for discussion/decision](#12-open-governance-points-for-discussiondecision)

13. [Appendix A - Key positions](#13-appendix-a---key-positions)    

## 1. Scope and Purpose

This guideline defines lifecycle management of Identity and Consent Management (ICM) versions and how CAMARA APIs declare and maintain compatibility with them. It defines under what rules API versions and ICM versions can co-evolve while still providing clear compatibility guarantees.

> The guideline recognizes operational reality: API Providers offer the same API version against multiple ICM versions, and API Consumers require an API version and an ICM version as part of a single usage contract — if either aspect changes in a way that affects them, their implementation must adapt.

Reviewer 2: ```suggestion
The guideline recognizes operational reality: API Providers offer the same API version against multiple ICM versions, and API Consumers require an API version and an ICM version as part of a single usage contract — if either aspect changes in a way that affects them, their implementation must adapt.

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119).
```
This change requires the words 'must', 'should', etc. to be capitalised throughout the document. This is just an editorial matter and does not affect content approval.

Author: Done

- **ICM-compatibility** (umbrella): both API version and API deployment ICM-compatibility together. Successful integration of an API Consumer with an API Provider requires both to hold.
- **ICM version**: a Semantic Versioning (SemVer 2.0) compliant version number with major, minor, and patch components, identifying a specific set of ICM artefacts. Starting with ICM 1.0.0, major-version increments indicate breaking changes for API definitions, API deployments, or both — changes that cannot be expressed additively.
- **ICM lifecycle states**: Supported / Deprecated / Retired / Revoked are the four possible lifecycle states of an ICM version (see [ICM version lifecycle states and governance](#5-icm-version-lifecycle-states-and-governance)). Lifecycle states apply per ICM version.
- **ICM governance**: the decision process to transition an ICM version to a different lifecycle state (see [ICM version lifecycle states and governance](#5-icm-version-lifecycle-states-and-governance)).

> - **Exception**: a time-bound, governance-approved authorization that permits a specific (API version, ICM version) pair outside the normal ICM-compatibility guarantee.

Reviewer 2: The glossary's definition of "Exception" only covers the ICM-compatibility exception type. Section 10 later introduces a second, structurally different exception type (the lifecycle-state exception toward Revoked), which is not reflected in the glossary entry. Readers who stop at section 2 for definitions will miss that a second exception mechanism exists.

Author: Good catch. Added the second exception type in the glossary in https://github.com/camaraproject/ReleaseManagement/commit/2d2bcbc11f7b3c756c9fbe1fe7ea42de61fe489c

  - each deployed API version must be ICM-compatible with the ICM version it operates against (design-time), or must be upgraded to a newer ICM-compatible API version;
  - the API Provider must implement an ICM version that allows ICM-compatibility of all deployed API versions (runtime).
- The API Consumer is responsible for ensuring that its own implementation matches the ICM version deployed by the API Provider, as well as the deployed API version.

A new ICM version may be due to changes in either or both classes of ICM information:

- **A change to ICM design info** (for example, a new scope format) → the API definition may need updating → triggers a new API version for impacted APIs (see [Minimum ICM version](#61-minimum-icm-version)). API Consumers consuming the new API version must adapt their requests to the new format.
- **A change to ICM deployment info** (for example, tightening assertion lifetime) → API Provider deployments and API Consumer implementations must update → does NOT by itself imply a new API version. The API definition is unchanged; only the runtime behavior changes.

> - **An ICM change to both** → both consequences apply.

Reviewer 2: ```suggestion
A new ICM version may result from a change to either or both classes of ICM information:
 
| Type of change | Consequence for the API definition | Who must act |
|---|---|---|
| Change to **ICM design info** (e.g. a new scope format) | Triggers a new API version for impacted APIs (see [Minimum ICM version](#61-minimum-icm-version)) | API designers release the new version; Consumers adapt requests to the new format |
| Change to **ICM deployment info** (e.g. tightening assertion lifetime) | No new API version required — the API definition is unchanged; only runtime behavior changes | Providers and Consumers update their implementations |
| Change to **both** | Both consequences apply | All three roles act |
```
It would look as follows. Feel free to ignore the proposal if you don't like it :)

--

A new ICM version may result from a change to either or both classes of ICM information:
 
| Type of change | Consequence for the API definition | Who must act |
|---|---|---|
| Change to **ICM design info** (e.g. a new scope format) | Triggers a new API version for impacted APIs (see [Minimum ICM version](#61-minimum-icm-version)) | API designers release the new version; Consumers adapt requests to the new format |
| Change to **ICM deployment info** (e.g. tightening assertion lifetime) | No new API version required — the API definition is unchanged; only runtime behavior changes | Providers and Consumers update their implementations |
| Change to **both** | Both consequences apply | All three roles act |

**Actions to maintain ICM-compatibility**

| meta-release | ICM version | ICM lifecycle state | API designer | API Provider | API Consumer |
|---|---|---|---|---|---|
| Signal N | vX.0.0 | Supported (24 months) |   | Deploy ICM vX.0.0 |   |

> |   | Older ICM version | Deprecated (12 months |   | Plan upgrade of APIs dependent on ICM versions in Deprecated state | Plan upgrade of API clients dependent on ICM versions in Deprecated state |

Reviewer 2: ```suggestion
|   | Older ICM version | Deprecated (12 months) |   | Plan upgrade of APIs dependent on ICM versions in Deprecated state | Plan upgrade of API clients dependent on ICM versions in Deprecated state |
```

Author: No longer applicable with new table.

|   | Older ICM version | Retired |   | Retire APIs dependent on ICM versions in Retired state | Retire API clients dependent on ICM versions in Retired state |
| Sync N |   |   | **New, initial and new stable APIs**: MUST release against Signal N. **Stable APIs**: If 2 previous Signal releases were skipped, API MUST release against Signal N. **Major update** →  MUST release against Signal N. **Minor or patch update** →  MAY skip Signal N, but maximally 2 subsequent Signal releases can be skipped (independent or Sync track) | Deploy Sync N APIs based on ICM vX.0.0 | Deploy Sync N API clients based on ICM vX.0.0 |
| Signal N+1 | vX+1.0.0 | Supported (24 months) |   | Deploy ICM vX+1.0.0 |   |

> |   | Older ICM version | Deprecated (12 months |   | Plan upgrade of APIs dependent on ICM versions in Deprecated state | Plan upgrade of API clients dependent on ICM versions in Deprecated state |

Reviewer 2: ```suggestion
|   | Older ICM version | Deprecated (12 months) |   | Plan upgrade of APIs dependent on ICM versions in Deprecated state | Plan upgrade of API clients dependent on ICM versions in Deprecated state |
```

Author: No longer applicable with new table.

| Signal N+3 | vX+3.0.0 | Supported (24 months) |   | Deploy ICM vX+3.0.0 |   |
|   | vX+1.0.0 | Deprecated (12 months) | Plan upgrade of APIs dependent on ICM vX+1.0.0 | Plan upgrade of API clients dependent on ICM vX+1.0.0 |   |
|   | vX.0.0 | Retired |   | Retire APIs dependent on ICM vX.0.0; Retire ICM vX.0.0 | Retire API clients dependent on ICM vX.0.0 |
| Sync N+3 |   |   | **New, initial and new stable APIs**: MUST release against Signal N+3. **Stable APIs**: If 2 previous Signal releases were skipped, API MUST release against Signal N+3. **Major update** →  MUST release against Signal N+3. **Minor or patch update** →  MAY skip Signal N+3, but maximally 2 subsequent Signal releases can be skipped (independent or Sync track)  | Deploy Sync N+3 APIs based on ICM vX+3.0.0 | Deploy Sync N+3 API clients based on ICM vX+3.0.0 |

> 

Reviewer 2: We could collapse the seven-row, five-column table repeating the same logic four times (once per Signal/Sync pair) into a single generic four-row table plus the Sync release rules as a separate list. The pattern is identical at every cycle, so repeating it three times added length without adding information.

```markdown
The table below defines the single repeating pattern of actions that applies at **every** Signal/Sync pair. It replaces a cycle-by-cycle listing: substitute `N`, `N+1`, `N+2`, … for any actual meta-release, and the same obligations apply each time a new major ICM version is released. A worked multi-year example is given in [section 8.3](#83-example-of-icm-lifecycle-state-evolution-and-api-deployment-across-meta-releases).
 
| Meta-release | Event | API designer | API Provider | API Consumer |
|---|---|---|---|---|
| **Signal N** | New major ICM version `vX.0.0` enters **Supported** state | — | Deploy ICM `vX.0.0` | — |
| | Older ICM version enters **Deprecated** state (12-month migration window) | — | Plan upgrade of APIs that depend on the Deprecated ICM version | Plan upgrade of API clients that depend on the Deprecated ICM version |
| | Older ICM version enters **Retired** state | — | Retire APIs that depend on the Retired ICM version | Retire API clients that depend on the Retired ICM version |
| **Sync N** | New API version release | Release against Signal N per the rules below\* | Deploy Sync N APIs based on ICM `vX.0.0` | Deploy Sync N API clients based on ICM `vX.0.0` |
 
\* **API release rules at Sync N:**
- New, initial, and new stable APIs **MUST** release against Signal N.
- A stable API **MUST** release against Signal N if it has already skipped the two previous Signal releases.
- A **major** API update **MUST** release against Signal N.
- A **minor or patch** API update **MAY** skip Signal N, but no more than two consecutive Signal releases may be skipped (whether on the independent or Sync track).
 
At any given time, at most two major ICM versions are relevant to a Provider's planning: the newly Supported one, and the one entering Deprecated (with a third, entering Retired, dropping out of scope). The table's four rows are the complete set of obligations that recur every cycle — nothing new is introduced at Signal N+1, N+2, or beyond; only the version numbers shift by one.
```

It would look as follows. Feel free to ignore the proposal if you don't like it :)

--

The table below defines the single repeating pattern of actions that applies at **every** Signal/Sync pair. It replaces a cycle-by-cycle listing: substitute `N`, `N+1`, `N+2`, … for any actual meta-release, and the same obligations apply each time a new major ICM version is released. A worked multi-year example is given in [section 8.3](#83-example-of-icm-lifecycle-state-evolution-and-api-deployment-across-meta-releases).
 
| Meta-release | Event | API designer | API Provider | API Consumer |
|---|---|---|---|---|
| **Signal N** | New major ICM version `vX.0.0` enters **Supported** state | — | Deploy ICM `vX.0.0` | — |
| | Older ICM version enters **Deprecated** state (12-month migration window) | — | Plan upgrade of APIs that depend on the Deprecated ICM version | Plan upgrade of API clients that depend on the Deprecated ICM version |
| | Older ICM version enters **Retired** state | — | Retire APIs that depend on the Retired ICM version | Retire API clients that depend on the Retired ICM version |
| **Sync N** | New API version release | Release against Signal N per the rules below\* | Deploy Sync N APIs based on ICM `vX.0.0` | Deploy Sync N API clients based on ICM `vX.0.0` |
 
\* **API release rules at Sync N:**
- New, initial, and new stable APIs **MUST** release against Signal N.
- A stable API **MUST** release against Signal N if it has already skipped the two previous Signal releases.
- A **major** API update **MUST** release against Signal N.
- A **minor or patch** API update **MAY** skip Signal N, but no more than two consecutive Signal releases may be skipped (whether on the independent or Sync track).
 
At any given time, at most two major ICM versions are relevant to a Provider's planning: the newly Supported one, and the one entering Deprecated (with a third, entering Retired, dropping out of scope). The table's four rows are the complete set of obligations that recur every cycle — nothing new is introduced at Signal N+1, N+2, or beyond; only the version numbers shift by one.

Author: great ! proposal ! done in https://github.com/camaraproject/ReleaseManagement/commit/d8748e67a8702929d67dd246157f5c54c280b4ae
included a few textual changes as well:
- we no longer need refereces to N, N+1, N+2, etc.
- API release rules in bullets 3 and 4 apply to stable APIs

### 3.4 Path from ICM through Commonalities into API definitions

ICM design info changes do not reach API definitions directly. The path is:

1. ICM defines or updates design info (security schemas, scope format, mandatory text, etc.).
2. Commonalities updates the API Design Guide (and other relevant artifacts) to mandate the new design info for API definitions.
3. API Sub Projects update their API versions to align with the new Commonalities API Design Guide.

This means an ICM design info change typically also triggers a Commonalities update. How tightly ICM and Commonalities lifecycles must be coupled is an open item (see [Open governance points for discussion/decision](#12-open-governance-points-for-discussiondecision)).

## 4. ICM versioning

This guideline assumes ICM adopts strict SemVer versioning starting with its 1.0.0 release. 

> An change in ICM version reflects all changes made for that version. These changes may concern either or both classes of ICM information that impact ICM-compatibility: ICM design info and ICM deployment info ([ICM-compatibility](#3-icm-compatibility)):

Reviewer 2: ```suggestion
A change in ICM version reflects all changes made for that version. These changes may concern either or both classes of ICM information that impact ICM-compatibility: ICM design info and ICM deployment info ([ICM-compatibility](#3-icm-compatibility)):
```

Author: Fixed

- changes that break **API version ICM-compatibility** — affecting ICM design info (scope format, `securitySchemes` syntax, schemas, operations), requiring API versions to be updated;
- changes that break **API deployment ICM-compatibility** — affecting ICM deployment info (auth flows, grant types, assertion format and lifetime, token processing), requiring API Provider and Consumer implementations to be updated, independently of the API version.

The two classes of changes are not necessarily related. For example, a change in ICM deployment info MAY (1) or MAY NOT (2) lead to a change of API design info, e.g.

> 1. a new auth flow impacts the API design info as a new type of credential/token needs to be introduced in the API definition (change of ICM deployment info leads to change of ICM design info)

Reviewer 2: ```suggestion
1. a new security scheme impacts the API design info as a new `securitySchemes` needs to be introduced in the API definition (change of ICM deployment info leads to change of ICM design info)
```
This is not accurate. In general, a new OAuth flow introduced by ICM does not affect the API specification. To provide a better example, we could say that ICM introduces new security scheme, changes the mandatory ICM info.description template or changes the scope format as mentioned in other parts of the document.

Author: Corrected

Author: @jpengar One question still on this one: the intention of the first example was to show that an ICM deployment info change may impact ICM design info. Would the following **additions in bold** be correct ? if not I will just leave it as is.

1. a new security scheme, **required by a new auth flow or by a new behaviour of an existing auth flow,**  impacts the API design info as a new `securitySchemes` **item** needs to be introduced in the API definition (change in ICM deployment info leads to change in ICM design info)

Reviewer 2: LGTM

The published lifecycle state must be available in machine-readable form for the CAMARA validation support to consume ([CAMARA validation support](#65-camara-validation-support)). Until a schema is defined, the ICM version release notes are the single authoritative source — readable by humans but not by tooling.

**ICM release notes - lifecycle state table**

The following table tracks the lifecycle state of ICM versions. This table is part of the ICM release notes.

Note: the entries in _italics_ are examples that should not be present in the release notes.

| ICM version | Lifecycle State | Supported | Deprecated | Retired | Revoked | Extension | Comment |
|---|---|---|---|---|---|---|---|
| _2.1.0_ | _Supported_ | _2028-08_ | _2030-08_ | _2031-04_ |  |  | _replacement of 2.0.0 due to security issue_ |
| _2.0.0_ | _Revoked_ | _2028-04_ | _2030-04_ | _2031-04_ | _2028-08_ |  | _override example_ |
| _1.0.0_ | _Supported_ | _2027-04_ | _2029-04_ | _2030-04_ |  |  | _future release example_ |

> | **0.5.0** | **Supported** | 2026-05 | 2028-04 | 2029-04 |  | _2027-12_ | new durations: 24 + 12 months; _Dextension eprecated + 6 months_ |

Reviewer 2: ```suggestion
| **0.5.0** | **Supported** | 2026-04 | 2028-04 | 2029-04 |  | _2027-12_ | new durations: 24 + 12 months; _extension deprecated + 6 months_ |
```
There is a typo in this row, did you mean "extension deprecated"?

Author: It  was ment to say "Deprecated extension" (short for extension of the Deprecated state duration)- fixed this.

the date you proposed to change was actually correct as the M4 of Spring26 was in may (05), but the example then aligns the Supported period to be 23 months instead of 24 to align back to the future Signal28 planned date. This is debateble of course and would be up to ICM to decide and document in the release notes.

I kept this date as 05

| 0.4.0 (r3.3) | info.description | Modification of the mandatory privacy legal text block and onboarding instructions. | Replace the current description block with the new official ICM v0.4.0 template. | `camara-icm-mandatory-text` |

**Table B — ICM deployment info - breaking changes** (impacting API deployment ICM-compatibility; see [API deployment ICM-compatibility](#32-api-deployment-icm-compatibility)). Used by API Providers and API Consumers to plan deployment updates.

| ICM version (tag) | Protocol Layer / Component | Affected runtime behavior | Action for API Providers | Action for API Consumers |
|---|---|---|---|---|
| _Semantic version and associated release tag (e.g., 0.3.0 (r2.3))_ | _The affected flow or endpoint (e.g., OIDC Discovery, Token Endpoint, CIBA Backchannel)_ | _The new strict validation or behavioral rule being introduced_ | _What the providers must configure in their Auth Server/infrastructure_ | _What the application developer must update in their token-request logic_ |

> | 0.3.0 (r2.3) | All Auth flows | Strict capping of the `private_key_jwt assertion` lifetime to a maximum of 300 seconds | The request SHALL be rejected by the authorisation server if the exp claim is more than 300 seconds later than the time of receipt. Additionally, if the iat claim is present, the request SHALL be rejected if the difference between the exp claim and iat claim is more than 300 seconds | The API Consumer MUST NOT create client assertions with a lifetime of more than 300 seconds, calculated as the difference between the exp (expires at) claim and the token creation time (which SHALL also be the value of the iat claim if present) |

Reviewer 2: ```suggestion
| 0.3.0 (r2.3) | All Auth flows | Strict capping of the `private_key_jwt` assertion lifetime to a maximum of 300 seconds | The request SHALL be rejected by the authorisation server if the exp claim is more than 300 seconds later than the time of receipt. Additionally, if the iat claim is present, the request SHALL be rejected if the difference between the exp claim and iat claim is more than 300 seconds | The API Consumer MUST NOT create client assertions with a lifetime of more than 300 seconds, calculated as the difference between the exp (expires at) claim and the token creation time (which SHALL also be the value of the iat claim if present) |
```

Author: Fixed

  - API deployment ICM-compatibility — affecting ICM deployment info (auth flows, grant types, assertion format and lifetime, token processing): No need to release a new API version. Maintaining API deployment ICM-compatibility is the responsibility of API Providers and API Consumers.

### 6.4 Determining the `x-camara-min-icm` value

At each API version public release (new major, minor, or patch), the API Sub Project determines the value using the following formula:

```
x-camara-min-icm = max (
  lowest Supported ICM version at API version public release,
  lowest Supported ICM version containing all features this API's definition requires,
  lowest ICM version required by the Commonalities version declared in `x-camara-commonalities`
)
```

Example:

- lowest Supported ICM version at API version public release: v1.0.0
- lowest Supported ICM version containing all features this API's definition requires: v2.1.0
- lowest ICM version required by the Commonalities version declared in `x-camara-commonalities`: v2.0.0

-> **x-camara-min-icm: v2.1.0**  (== max (v1.0.0, v2.1.0, v2.0.0))

> For APIs with no ICM-version-specific feature dependencies beyond what Commonalities mandates, the second element is not applicable. This reduces the formula to `max (lowest Supported ICM version at API version public release, lowest ICM version required by the Commonalities version declared in `x-camara-commonalities`). 

Reviewer 2: ```suggestion
For APIs with no ICM-version-specific feature dependencies beyond what Commonalities mandates, the second element is not applicable. This reduces the formula to `max (lowest Supported ICM version at API version public release, lowest ICM version required by the Commonalities version declared in x-camara-commonalities)`. 
```

Author: Fixed

### 8.2 Out-of-cycle ICM releases

ICM releases outside of the Signal meta-release are allowed and sometimes required, e.g., for security vulnerabilities patches, defect corrections, or urgent regulatory changes. These changes impact the ICM version as per SemVer guidelines. 

ICM version lifecycle state transitions may occur off-cycle in security-driven cases, or on explicit governance decision or high-priority TSC-driven updates. 

### 8.3 Example of ICM lifecycle state evolution and API deployment across meta-releases

This section illustrates a typical scenario starting from Signal27, based on the ICM lifecycle state durations of 24 months (2 years) Supported + 12 months (1 year) Deprecated = 36 months (3 years) total.

- [**ICM Release in 2027**] Q2 2027: Signal27 released -> Q4 2027: Sync27 APIs released.
- [**API Deployment**] Q2 2028: An operator launches Sync27 APIs in production (utilizing their 2028 budget).
- [**ICM Release in 2028**] Q2 2028: Signal28 released -> Q4 2028: Sync28 APIs released.
- [**ICM Release in 2029**] Q2 2029: Signal29 released -> Q4 2029: Sync29 APIs released.

> - [**ICM Deprecation**] Q2 2040: Signal27 becomes Deprecated (24 months after Signal28 was released in Q2 2028).

Reviewer 2: ```suggestion
- [**ICM Deprecation**] Q2 2030: Signal27 becomes Deprecated (24 months after Signal28 was released in Q2 2028).
```

Author: Fixed, but this part has been rewritten

- [**ICM Release in 2030**] Q2 2030: Signal30 released -> Q4 2030: Sync30 APIs released.

> - [**ICM Retirement**] Q2 2031: Signal27 becomes Retired (36 onths after Signal28 was released).

Reviewer 2: ```suggestion
- [**ICM Retirement**] Q2 2031: Signal27 becomes Retired (36 months after Signal28 was released).
```

Author: Fixed, but this part has been rewritten

- Security Scheme Object - `securitySchemes` field [(ref)](https://github.com/camaraproject/IdentityAndConsentManagement/blob/r3.3/documentation/CAMARA-API-access-and-user-consent.md#use-of-openidconnect-for-securityschemes), and related schemas
- Scope format - `scopes` field [(ref)](https://github.com/camaraproject/Commonalities/blob/main/documentation/CAMARA-API-Design-Guide.md#66-scope-naming), and related schemas. Please note that ICM also defines how to declare the [purpose as a scope](https://github.com/camaraproject/IdentityAndConsentManagement/blob/r3.3/documentation/CAMARA-Security-Interoperability.md#purpose-as-a-scope), but this does not affect the API definitions. This only applies to authentication requests.

**ICM deployment info**

This is the information defined by an ICM version that applies to API deployments (including at API Provider/Consumer design time and at runtime) — authentication flows, grant types, assertion format and lifetime, token processing, claim handling, and similar behaviors between API Provider and Consumer. 

> ### 3.2 API version ICM-compatibility

Reviewer 2: @tanjadegroot take a look at the index and the numbers of the subsections. They should start from 3.1, but currently start from 3.2. Consequently, the links in the index are broken.

Reviewer 3: @tanjadegroot I will take an update to make the numbering and links consistent

Reviewer 3: Done in https://github.com/camaraproject/ReleaseManagement/pull/545/commits/451a41de0f791cb8ad7db92c86b0340f75abd31d together with some typos.
